rfd/0128-autogenerate-resource-reference.md

```diff
@@ -0,0 +1,600 @@
+
+However, we can't guarantee that every Teleport resource available to `tctl`
+will correspond to a protobuf message definition. Since we generate Go source
+files based on our `.proto` files, and Teleport needs to unmarshal all Teleport
+resources into structs from YAML, it's a safe bet that all Teleport resources
+will have a corresponding Go struct.
+
+The downside of using Go source files is that, if we want to change the
+description of a field or some other aspect of the generated reference, we need
+to: 
+
+1. Locate the Go struct that corresponds to an entry within the reference page.
+1. Determine if the Go struct was generated from a `proto` file and, based on
+   this, edit either the Go struct definition or the underlying `proto` file.
+1. If relevant, generate a Go source file from the edited protobuf message
+   definition.
+1. Regenerate the reference page.
+
+I don't imagine this will be prohibitive, though, as long as we:
+
+- Indicate within the generated reference page which Go source file each entry
+  is based on.
+- Edit the comments in Go type definitions and `proto` files to ensure
+  consistency, clarity, and grammatical correctness.
+
+Having this Make target depend on `make grpc` will also automate this flow.
+
+### The output
+
+The output of the program will be a Teleport resource reference generated
+from a template.
+
+#### Template format and data
+
+The generator will produce a template based on a `Resource` struct with the
+following fields:
+
+```go
+type Resource struct {
+  SectionName string
+  Description string
+  SourcePath  string
+  Fields      []Field
+  YAMLExample string
+}
+
+type Field struct{
+  Name        string
+  Description string
+  Type        string
+}
+```
+
+The template will look similar to this, anticipating a `[]Resource`:
+
+```text
+{{ range . }}
+## {{ .SectionName }}
+
+{{ .Description }}
+
+{/*Automatically generated from: {{ .SourcePath}}*/}
+
+|Field Name|Description|Type|
+|---|---|---|
+{{ range .Fields }}
+|.Name|.Description|.Type|
+{{ end }} 
+
+{{ .YAMLExample }}
+{{ end }}
+```
+
+#### Template example
+
+Here is an example of a populated template for an application (`AppSpecV3` in
+`api/types/types.pb.go`): 
+
+```markdown
+## App Spec V3
+
+App Spec V3 is the specification for an application registered with Teleport.
+
+{/*Automatically generated from: api/types/types.pb.go*/}
+
+|Field Name|Description|Type|
+|---|---|---|
+|`uri`|The web app endpoint.|`string`|
+|`public_addr`|The public address the application is accessible at.|`string`|
+|`dynamic_labels`|The app's command labels.|`map[string]`[Command LabelV2](#command-label-v2)|
+|`insecure_skip_verify`|Disables app's TLS certificate verification.|`boolean`|
+|`rewrite`|A list of rewriting rules to apply to requests and responses.|[Rewrite](#rewrite)|
+|`aws`|Contains additional options for AWS applications.|[App AWS](#app-aws)|
+|`cloud`|Identifies the cloud instance the app represents.|`string`|
+
```

Comment:
Should the template create a header or other copy to specify that this is an excerpt from a specific config file and/or resource definition?

Reply:
Do you mean that, if a type is a field within a resource definition, the reference for the field type should list the resource definitions it's part of?

Reply:
I'm pretty sure this comment refers to something the template includes already—the source of each `tctl`-applicable resource—so I'm going to go ahead and merge this. Feel free to open a PR/issue if there's anything I should change!